File: day12/main.lua

File = "./example.txt"
io.input(File)

function read_input()
	local map = {}
	for line in io.lines() do
		local row = {}
		for pot in line:gmatch('.') do
			row[#row+1] = pot
		end
		map[#map+1] = row
	end
	return map
end

function check_neighbors(x, y, map)
	for i=-1, 1 do
		for j = -1,1 do
			if math.abs(i) ~= math.abs(j) then
				local neighbor = map[x +i][y+j]
				local pot = map[x][y]
				if pot == neighbor then return true end
			end
		end
	end
	return false
end

function area(map, area_tbl)
	for i=2, #map-1 do
		for j=2, #map[i]-1 do
			if check_neighbors(i, j, map) then
				area_tbl[map[i][j]] = area_tbl[map[i][j]] + 1
			end
		end
	end
end

function pad_map(map)
	local x = #map+2
	local y = #map[1]+2
	local padded_map = {}
	for i=1,x do
		local row = {}
		for j=1,y do
			row[#row+1] = "."
		end
		padded_map[#padded_map+1] = row
	end
	for i, row in ipairs(map) do
		for j, val in ipairs(row) do
			padded_map[i+1][j+1] = val
		end
	end
	return padded_map
end

function recursive_neighbours(x, y, pot, map, map_copy, region_tbl)
	for i=-1, 1 do
		for j = -1,1 do
			if math.abs(i) ~= math.abs(j) then
				local neighbor = map[x +i][y+j]
			end
		end
	end
end

function count_perimeter(x, y, pot, map, perimeter_tbl)
	for i=-1, 1 do
		for j = -1,1 do
			if math.abs(i) ~= math.abs(j) then
				local neighbor = map[x +i][y+j]
				if neighbor ~= pot then
					perimeter_tbl[pot] = perimeter_tbl[pot] + 1
				end
			end
		end
	end
end

function get_region_map( pot, map )
	local new_map = {}
	for i=2, #map-1 do
		local new_row = {}
		for j=2, #map[i]-1 do
			if map[i][j] == pot then
				new_row[#new_row+1] = "#"
			else
				new_row[#new_row+1] = "."
			end
		end
		new_map[#new_map+1] = new_row
	end
	return new_map
end

function p1()
	local map = read_input()
	local perimeter_tbl = {}
	local area_tbl= {}
	local pots = {}
	for i, row in ipairs(map) do
		for j, val in ipairs(row) do
			if perimeter_tbl[val] == nil then
				perimeter_tbl[val] = 0
				area_tbl[val] = 0
			end
		end
	end
	for k, v in pairs(area_tbl) do
		pots[#pots+1] = k
	end
	local padded_map = pad_map(map)
	for _, pot in ipairs(pots) do
		local region_map = get_region_map(pot, padded_map)
	end
end

p1()

File: day12/main.py

def get_input(file):
    garden = []
    with open(file, 'r') as f:
        map = f.readlines()
        for row in map:
            new_row = []
            for pot in row:
                if pot != "\n":
                    new_row.append(pot)
            garden.append(new_row)
    return garden


def recursion(pot, x, y, garden, visited=set()):
    m = len(garden) -1
    n = len(garden[0]) -1
    if (x,y) in visited or x > m or x < 0 or y > n or y < 0: 
        return visited
    elif garden[x][y] == pot:
        visited.add((x,y))
        visited = recursion(pot, x, y+1, garden, visited)
        visited = recursion(pot, x, y-1, garden, visited)
        visited = recursion(pot, x+1, y, garden, visited)
        visited = recursion(pot, x-1, y, garden, visited)

    return visited
    
def recursionp2(pot, x, y, garden, new_map, visited=set()):
    m = len(garden) -1
    n = len(garden[0]) -1

    if (x,y) in visited or x > m or x < 0 or y > n or y < 0: 
        if (x,y) not in visited:
            new_map[x+1][y+1] = "*"
        return visited, new_map
    elif garden[x][y] == pot:
        visited.add((x,y))
        new_map[x+1][y+1] = pot
        visited, new_map = recursionp2(pot, x, y+1, garden, new_map, visited)
        visited, new_map = recursionp2(pot, x, y-1, garden, new_map, visited)
        visited, new_map = recursionp2(pot, x+1, y, garden, new_map, visited)
        visited, new_map = recursionp2(pot, x-1, y, garden, new_map, visited)
    else:
        new_map[x+1][y+1] = "*"

    return visited, new_map

def p1():
    garden = get_input('./input.txt')
    pots = []
    for row in garden:
        for pot in row:
            if pot not in pots:
                pots.append(pot)
    total = 0
    for pot in pots:
        total_visited = set()
        fence = 0
        visited = set()
        for x, row in enumerate(garden):
            for y, val in enumerate(row):
                if val == pot and (x, y) not in total_visited:
                    visited = recursion(pot, x, y, garden, visited)
                    for i, j in visited:
                        total_visited.add((i,j))
                        for m in range(-1, 2):
                            for n in range(-1,2):
                                if abs(m) == abs(n):
                                    continue
                                if (i+m, j+n) not in visited:
                                    fence +=1
                            
                    total += len(visited)*fence
                    print(pot, len(visited), fence)
                else:
                    fence =0
                    visited = set()

    print(total)


            
def p2():
    garden = get_input('./example.txt')
    pots = []

    for row in garden:
        for pot in row:
            if pot not in pots:
                pots.append(pot)
    total = 0
    for pot in pots:
        total_visited = set()
        fence = 0
        visited = set()
        region_map = []
        for i in range(len(garden)*2):
            new_row = []
            for j in range(len(garden[0])*2):
                new_row.append("")
            region_map.append(new_row)
        for x, row in enumerate(garden):
            for y, val in enumerate(row):
                if val == pot and (x, y) not in total_visited:
                    visited, region_map = recursionp2(pot, x, y, garden, region_map, visited)
                    for row in region_map:
                        print(row)
                    sides = 0
                    for i, j in visited:
                        total_visited.add((i,j))
                        for k, n in visited:
                            if i != k and j != n:
                                for m in range(-1, 2):
                                    for n in range(-1,2):
                                        if abs(m) != abs(n) and m !=0:
                                            if (i+m, j+n) not in visited:
                                                fence+=1
                            
                    total += len(visited)*fence
                    print(pot, len(visited), fence)

                    #print(direction_count)
                else:
                    fence =0
                    visited = set()

    print(total)

if __name__ == "__main__":
    p2()
